rig/machine_control/boot.py

```diff
@@ -0,0 +1,160 @@
+    height : int
+        Height of the machine (0 < h < 256)
+    cpu_frequency : int
+        CPU clock-frequency.  **Note**: The default (200 MHz) is known
+        safe.
+    hardware_version : int
+    led_config : int
+    boot_data : bytes
+        Data to boot the machine with.
+    struct_data : bytes or None
+        Data to interpret as a representation of the memory layout of data
+        within the boot data.
+
+    Notes
+    -----
+    The constants `rig.boot.spinX_boot_options` can be used to specify boot
+    parameters, for example:
+
+        boot("board1", **spin5_boot_options)
+
+    Will boot the Spin5 board connected with hostname "board1".
+    """
+    # Read the struct file and modify the "sv" struct to contain the
+    # configuration values and write this into the boot data.
+    sv = struct_file.read_struct_file(struct_data)[b"sv"]
+    sv.update_default_values(p2p_dims=(width << 8) | height,
+                             hw_ver=hardware_version,
+                             cpu_clk=cpu_frequency,
+                             led0=led_config,
+                             unix_time=int(time.time()),
+                             boot_sig=int(time.time()),
+                             root_chip=1)
+    struct_packed = sv.pack()
+
+    # NOTE I am assured by ST that writing into boot data bytes 384 onwards
+    # is unlikely to change in the foreseeable future.
+    buf = bytearray(boot_data)
+    buf[BOOT_DATA_OFFSET:BOOT_DATA_OFFSET+BOOT_DATA_LENGTH] = \
+        struct_packed[:BOOT_DATA_LENGTH]
+    assert len(buf) < 32768  # Assert that we fit in DTCM
+    boot_data = bytes(buf)
+
+    # Create a socket to communicate with the board
+    sock = socket.socket(type=socket.SOCK_DGRAM)
+    sock.connect((hostname, PORT))
+
+    # Transmit the boot data as a series of SDP packets.  First determine
+    # how many blocks must be sent and transmit that, then transmit each
+    # block.
+    n_blocks = len(buf) // BOOT_BYTE_SIZE + (
+        1 if len(buf) % BOOT_BYTE_SIZE != 0 else 0)
+    assert n_blocks <= BOOT_MAX_BLOCKS
+
+    boot_packet(sock, 1, arg3=n_blocks - 1)
+    time.sleep(BOOT_DELAY)
+
+    block = 0
+    while len(boot_data) > 0:
+        # Get the data to transmit
+        data, boot_data = (boot_data[:BOOT_BYTE_SIZE],
+                           boot_data[BOOT_BYTE_SIZE:])
+
+        # Transmit, delay and increment the block count
+        a1 = ((BOOT_WORD_SIZE - 1) << 8) | block
+        boot_packet(sock, 3, a1, data=data)
+        time.sleep(BOOT_DELAY)
+        block += 1
+
+    # Send the END command
+    boot_packet(sock, 5, 1)
+
+    # Close the socket and give time to boot
+    sock.close()
+    time.sleep(2.0)
+
+
+def boot_packet(sock, cmd, arg1=0, arg2=0, arg3=0, data=b""):
+    """Create and transmit a packet to boot the machine.
+
+    Parameters
+    ----------
+    sock : :py:class:`~socket.socket`
+        Connected socket to use to transmit the packet.
+    cmd : int
```

Comment:
Can these be named/Enum'd?

    class BootCommand(IntEnum):
        """Boot packet command numbers"""
        
        start = 1
        send_block = 3
        finish = 5